turbopack/crates/turbopack-core/src/resolve/mod.rs

```diff
@@ -227,34 +247,36 @@ impl ModuleResolveResult {
                 .copied(),
         );
     }
-
-    pub fn is_unresolvable_ref(&self) -> bool {
-        self.primary.is_empty()
-    }
 }
 
 #[turbo_tasks::value_impl]
 impl ModuleResolveResult {
     #[turbo_tasks::function]
     pub async fn with_affecting_source(
-        self: Vc<Self>,
+        &self,
         source: ResolvedVc<Box<dyn Source>>,
     ) -> Result<Vc<Self>> {
-        let mut this = self.owned().await?;
-        this.add_affecting_source_ref(source);
-        Ok(this.into())
+        let mut affecting_sources = self.affecting_sources.to_vec();
+        affecting_sources.push(source);
```

Comment:
This is doing 3 allocations (to_vec of size N, push of size 2N, into_boxed_slice of size N + 1).
It could do 1 when creating a Vec with the correct capacity

Reply:
IMO, the best way to do this is with iterators and collect. `Iterator::size_hint` provides the right information for the `Vec` to be correctly sized: [https://doc.rust-lang.org/nightly/core/iter/trait.ExactSizeIterator.html](https://doc.rust-lang.org/nightly/core/iter/trait.Iterator.html#method.size_hint)

You can also collect directly into a boxed slice (it creates a `Vec` under the hood first though, it's just a convenience implementation).

Something like this:

```
let affecting_sources: Box<[_]> = self.affecting_sources
    .iter()
    .copied()
    .chain(iter::once(source))
    .collect()
```